apiserver/fun/main.go

```diff
@@ -4,40 +4,164 @@ import (
 	"encoding/json"
 	"flag"
 	"fmt"
-	"io/ioutil"
-	"os"
 	"runtime"
 
 	"k8s.io/kubernetes/pkg/api"
 )
 
 var store map[int]*api.Pod
 
+// let's get the plain size of a Pod in memory
+// assume each field takes about 16 bytes additional space
+// unversioned.TypeMeta = 2 -> 32 bytes
+// ObjectMeta = 16 -> 256 bytes
+// PodSpec = 14 -> 224 bytes
+// PodStatus = 9 -> 144 bytes
+// Pod -> 4 -> 64 bytes
+// total = 45 -> 720 bytes
+
+// calculation time!
+// 100k empty pods!
+// 72105984 bytes = 721 bytes -> wow accurate
+
+// OK. Now let's fill in some fields to see the size explosion.
+// And figure out exactly which fields cause the pod grows from
+// 720 bytes to 3000 bytes...
+
+// 1.
+// Let's do Typemeta first. Additional values are 5 bytes
+// Ah... seems like there is no significant change as expected!
+
+// 2.
+// Let's do ObjectMeta then... Additional values are around 400 bytes.
+// So ideally, now each pod should take 1100 bytes.
+// However... It blows up... each pod now actually takes 2000 bytes.
+// So where are the 900 bytes come from?
+//
+// Sure. ObjectMeta contains some non-primitive types like map.
+// Is map the real issue? Let's ignore label map, which contains 15 bytes data.
+// Wow... Now each pod takes 1600 bytes... So each non-empty map takes 400 bytes!
+
+// Let's count how many maps does a non-empty pod has in the ObjectMeta -> 2!
+// OK. Now we learned that additional 900 bytes are from two stupid map structs in go.
+
+// 3.
+// Let's do podspec. Additional values are 80 bytes or so.
+// We expects the pod to take (720 + 80) = 800 bytes.
+// However, now the pod takes 1900 bytes.
+// Interesting... Where are the additional 1100 bytes?
+// Probably the stupid container array struct?
+// Ah... each container has a field called Resources.
+// Each Resources has two maps... Each map takes 400 bytes...
```

Comment:
I think maps are being created for every object, even those without resources / annotations / labels because of omitempty in JSON (which on deserialize will create an empty map).